File: others/string-concat/lua/benchmarker.lua
-- -*- coding: utf-8 -*-

---
--- $Release: $
--- $Copyright: copyright(c) 2011 kuwata-lab.com all rights reserved $
--- $License: MIT License $
---

--print(package.path)
socket = require('socket')
--table = require('table')
--print(socket._VERSION)


module('benchmarker', package.seeall)


Benchmarker = {}
Benchmarker.new = function(loop)
	local this = {
		loop        = loop or 100*1000,
		_tasks      = {},
		_empty_task = null,

		empty_task =
			function(self, body)
				self._empty_task = body
			end,

		task =
			function(self, title, body)
				table.insert(self._tasks, {title=title, body=body})
			end,

		run =
			function(self)
				self:print_environment()
				print('# loop = ' .. self.loop)
				local i, j
				local tasks = {}
				if self._empty_task then
					table.insert(tasks, {title='(Empty)', body=self._empty_task})
				end
				for i = 1, #self._tasks do
					table.insert(tasks, self._tasks[i])
				end
				local cycle = os.getenv('C') or 3
				for c = 1, cycle do
					print()
					self:_run_task(self.loop, c, tasks)
				end
			end,

		_run_task =
			function(self, loop, cycle, tasks)
				print(string.format("%-35s %10s %10s", '# cycle='..cycle, 'real', 'actual'))
				empty_time = 0.0
				for j = 1, #tasks do
					local t = tasks[j]
					local start = socket.gettime()
					t.body(loop)
					local stop = socket.gettime()
					local real = stop - start
					local actual = real - empty_time
					if j == 1 and self._empty_task then
						empty_time = real
					end
					print(string.format("%-35s %10.3f %10.3f", t.title, real, actual))
				end
			end,

		print_environment =
			function(self)
				print(string.format("# %-20s %s", '_VERSION:', _VERSION))
			end,

	}
	return this
end

File: others/string-concat/php/Benchmarker.php
<?php

///
/// $Release: $
/// $Copyright: copyright(c) 2011 kuwata-lab.com all rights reserved $
/// $License: MIT License $
///


class Benchmarker {
    var $_empty_task = null;
    var $_tasks = array();

    function __construct($loop=100000) {
        $this->loop = $loop;
    }

    function emptyTask($body) {
        $this->_empty_task = $body;
    }

    function task($title, $body) {
        $this->_tasks[] = array($title, $body);
    }

    function run($loop=null) {
        $this->print_environment();
        $tasks = $this->_tasks;
        if ($this->_empty_task) {
            array_unshift($tasks, array("(Empty)", $this->_empty_task));
        }
        if (! $loop) $loop = $this->loop;
        print("# loop=$loop\n");
        $cycle = getenv('C') ? (int)getenv('C') : 3;
        for ($i = 1; $i <= $cycle; $i++) {
            $this->_run_tasks($loop, $i, $tasks);
        }
    }

    function _run_tasks($loop, $cycle, $tasks) {
        print("\n");
        printf("%-30s%10s%10s\n", "# cycle=$cycle", "real", "actual");
        $empty_time = 0.0;
        foreach ($tasks as $i=>$pair) {
            list($title, $body) = $pair;
            if (function_exists('gc_collect_cycles')) {
                gc_collect_cycles();
            }
            printf('%-30s', $title);
            $start = microtime();
            $body($loop);
            $stop = microtime();
            $real = $this->_elapsed($start, $stop);
            $actual = $real - $empty_time;
            if ($i == 0 && $this->_empty_task) {
                $empty_time = $real;
            }
            printf("%10.3f%10.3f\n", $real, $actual);
        }
    }

    function _elapsed($start, $stop) {
        list($msec1, $sec1) = explode(' ', $start);
        list($msec2, $sec2) = explode(' ', $stop);
        $sec  = intval($sec2) - intval($sec1);
        $msec = floatval($msec2) - floatval($msec1);
        return $sec + $msec;
    }

    function print_environment() {
        printf("# %-20s%s\n", 'PHP_VERSION:', PHP_VERSION);
        printf("# %-20s%s\n", 'PHP_OS:', PHP_OS);
        //printf("# %-20s%s\n", 'php_uname:', php_uname());
        //printf("# %-20s%s\n", 'loop:', $this->loop);
        print("\n");
    }

}

